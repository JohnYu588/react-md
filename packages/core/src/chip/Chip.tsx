"use client";
import { cnb } from "cnbuilder";
import {
  forwardRef,
  type ButtonHTMLAttributes,
  type CSSProperties,
  type HTMLAttributes,
  type ReactNode,
} from "react";
import { type BackgroundColor } from "../cssUtils.js";
import { getIcon } from "../icon/iconConfig.js";
import { useElementInteraction } from "../interaction/useElementInteraction.js";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren.js";
import { useMaxWidthTransition } from "../transition/useMaxWidthTransition.js";
import { type PropsWithRef } from "../types.js";
import { chip, chipContent } from "./styles.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-chip-solid-background-color"?: string;
    "--rmd-chip-solid-disabled-background-color"?: string;
    "--rmd-chip-solid-color"?: string;
    "--rmd-chip-theme-background-color"?: string;
    "--rmd-chip-theme-color"?: string;
    "--rmd-chip-outline-background-color"?: string;
    "--rmd-chip-outline-color"?: string;
  }
}

/**
 * @since 6.0.0 Renamed the `noninteractive` prop to
 * `noninteractable`.
 */
export interface ChipProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * @defaultValue `"solid"`
   */
  theme?: "outline" | "solid";

  /**
   * Set this to `true` if the chip should gain additional box shadow while the
   * user is pressing down on the chip.
   *
   * @defaultValue `false`
   */
  raisable?: boolean;

  /**
   * @defaultValue `false`
   */
  disabled?: boolean;

  /**
   * When this is a boolean, a selected icon will be displayed before the
   * `children` and appear/disappear when `true`/`false`.
   *
   * @see {@link selectedIcon}
   * @see {@link selectedIconAfter}
   * @see {@link selectedThemed}
   */
  selected?: boolean;

  /**
   * Set this to `true` if the chip should change background color while
   * {@link selected} instead of displaying an icon. The default background
   * color is a lighter swatch of the theme primary color.
   *
   * @defaultValue `false`
   */
  selectedThemed?: boolean;

  /**
   * An optional className to provide only while {@link selected} is `true`.
   */
  selectedClassName?: string;

  /**
   * This will be ignored if {@link selected} is `undefined`.
   *
   * @defaultValue `getIcon("selected")`
   */
  selectedIcon?: ReactNode;

  /**
   * Set this to `true` if the {@link selectedIcon} should display as the
   * {@link rightAddon} instead of the {@link leftAddon}.
   *
   * @defaultValue `false`
   * @since 6.0.0
   */
  selectedIconAfter?: boolean;

  /**
   * Set this to `true` if the {@link selectedIcon} should not animate and
   * instead apply `display: none`
   *
   * @defaultValue `false`
   */
  disableIconTransition?: boolean;

  /**
   * Set this to `true` to render the chip as a `<span>` instead of a button.
   *
   * @since 2.6.0
   * @since 6.0.0 Renamed from `noninteractable`
   * @defaultValue `false`
   */
  noninteractive?: boolean;

  /**
   * An optional background color theme to use.
   *
   * @since 6.0.0
   */
  backgroundColor?: BackgroundColor;

  /**
   * An optional icon, avatar, circular progress, or custom component to render
   * before the `children`. This will remove some of the leading horizontal
   * padding on the chip as well.
   */
  leftAddon?: ReactNode;

  /**
   * An optional icon, avatar, circular progress, or custom component to render
   * after the `children`. This will remove some of the trailing horizontal
   * padding on the chip as well.
   */
  rightAddon?: ReactNode;

  /**
   * Set this to true if the `children` should not be wrapped in a span to apply
   * the {@link chipContent} styles which allow for shrinking text and
   * truncating text with ellipsis.
   *
   * @defaultValue `false`
   */
  disableContentWrap?: boolean;

  /**
   * This will be ignored if {@link disableContentWrap} is `true`.
   *
   * @since 6.0.0
   */
  contentProps?: PropsWithRef<HTMLAttributes<HTMLSpanElement>, HTMLSpanElement>;

  /**
   * This will be ignored if {@link disableContentWrap} is `true`.
   */
  contentStyle?: CSSProperties;

  /**
   * This will be ignored if {@link disableContentWrap} is `true`.
   */
  contentClassName?: string;
}

/**
 * **Server Component**
 *
 * @example Simple Example
 * ```tsx
 * import { Chip } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return <Chip>Content</Chip>;
 * }
 * ```
 *
 * @example Outlined Example
 * ```tsx
 * import { Chip } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return <Chip theme="outline">Content</Chip>;
 * }
 * ```
 *
 * @example Addons Example
 * ```tsx
 * import { Avatar, Chip } from "@react-md/core";
 * import AddCircleIcon from "@react-md/material-icons/AddCircleIcon";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Chip
 *       leftAddon={
 *         <Avatar>
 *           <img src="https://i.pravatar.cc/40?img=3" alt="" />
 *         </Avatar>
 *       }
 *       rightAddon={<AddCircleIcon />}
 *     >
 *       Chip
 *     </Chip>
 *   );
 * }
 * ```
 *
 * @since 6.0.0 Renamed the `noninteractable` prop to
 * `noninteractive`.
 */
export const Chip = forwardRef<HTMLButtonElement, ChipProps>(
  function Chip(props, ref) {
    const {
      "aria-pressed": ariaPressed,
      theme = "solid",
      className,
      raisable = false,
      disabled = false,
      selected,
      selectedThemed = false,
      selectedClassName,
      selectedIcon: propSelectedIcon,
      selectedIconAfter = false,
      backgroundColor,
      noninteractive = false,
      disableIconTransition = false,
      children: propChildren,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      leftAddon: propLeftAddon,
      rightAddon: propRightAddon,
      contentProps,
      contentStyle,
      contentClassName,
      disableContentWrap = false,
      ...remaining
    } = props;

    let buttonProps: ButtonHTMLAttributes<HTMLButtonElement> | undefined;
    if (!noninteractive) {
      buttonProps = {
        "aria-pressed": ariaPressed ?? (!!selected || undefined),
        type: "button",
        disabled,
      };
    }

    let content = propChildren;
    if (!disableContentWrap) {
      content = (
        <span
          style={contentStyle}
          {...contentProps}
          className={cnb(
            chipContent({ className: contentClassName }),
            contentProps?.className
          )}
        >
          {propChildren}
        </span>
      );
    }

    let leftAddon = propLeftAddon;
    let rightAddon = propRightAddon;
    const selectedIconNode = getIcon("selected", propSelectedIcon);
    const isTransitionable =
      !selectedThemed &&
      typeof selected === "boolean" &&
      typeof (selectedIconAfter ? propRightAddon : propLeftAddon) ===
        "undefined";
    const selectedIcon = useMaxWidthTransition({
      element: selectedIconNode,
      transitionIn: !!selected,
      disabled: !isTransitionable,
      disableTransition: disableIconTransition,
    });
    if (isTransitionable) {
      if (selectedIconAfter) {
        rightAddon = selectedIcon;
      } else {
        leftAddon = selectedIcon;
      }
    }

    const { pressed, pressedClassName, ripples, handlers } =
      useElementInteraction({
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || noninteractive,
      });
    const children = useHigherContrastChildren(content);
    const Component = noninteractive ? "span" : "button";

    return (
      <Component
        {...remaining}
        {...buttonProps}
        {...handlers}
        className={chip({
          className,
          theme,
          pressed: raisable && pressed,
          disabled,
          selected,
          selectedThemed,
          selectedClassName,
          noninteractive,
          backgroundColor,
          pressedClassName,
          leftAddon: !!leftAddon && (selectedIconAfter || selected !== false),
          rightAddon:
            !!rightAddon && (!selectedIconAfter || selected !== false),
        })}
        ref={ref}
      >
        {leftAddon}
        {children}
        {rightAddon}
        {ripples}
      </Component>
    );
  }
);
