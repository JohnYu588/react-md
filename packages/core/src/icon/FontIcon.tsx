import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";

import type { FontIconClassNameOptions } from "./styles";
import { icon } from "./styles";

export interface FontIconProps
  extends Omit<HTMLAttributes<HTMLElement>, "color">,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 *
 * \@remarks \@since 6.0.0 Switched from `<i>` to `<span>` element.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(props, ref) {
    const {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      color,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...remaining
    } = props;

    return (
      <span
        {...remaining}
        aria-hidden={ariaHidden}
        ref={ref}
        className={icon({
          type: "font",
          dense,
          color,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </span>
    );
  }
);
