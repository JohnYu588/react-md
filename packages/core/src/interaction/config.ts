import { type ElementInteractionMode } from "./types.js";

/**
 * @remarks \@since 6.0.0
 */
export interface ElementInteractionConfiguration {
  mode: ElementInteractionMode;
  higherContrast: boolean;
}

// NOTE: Uses get/set for test mocking

let mode: ElementInteractionMode = "ripple";
let higherContrast = true;

/**
 * @remarks \@since 6.0.0
 */
export const INTERACTION_CONFIG: ElementInteractionConfiguration = {
  get mode() {
    return mode;
  },
  set mode(nextMode: ElementInteractionMode) {
    mode = nextMode;
  },
  get higherContrast() {
    return higherContrast;
  },
  set higherContrast(nextHigherContrast) {
    higherContrast = nextHigherContrast;
  },
};
