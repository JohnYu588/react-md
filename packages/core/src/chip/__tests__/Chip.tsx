import { describe, expect, it, jest } from "@jest/globals";
import { type ReactElement, createRef } from "react";

import { FontIcon } from "../../icon/FontIcon.js";
import { INTERACTION_CONFIG } from "../../interaction/config.js";
import {
  fireEvent,
  render,
  rmdRender,
  screen,
  userEvent,
} from "../../test-utils/index.js";
import { useToggle } from "../../useToggle.js";
import { DISPLAY_NONE_CLASS } from "../../utils/isElementVisible.js";
import { Chip } from "../Chip.js";

describe("Chip", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLButtonElement>();
    const props = {
      ref,
      children: "Chip",
    } as const;
    const { rerender } = render(<Chip {...props} />);

    const chip = screen.getByRole("button", { name: "Chip" });
    expect(ref.current).toBeInstanceOf(HTMLButtonElement);
    expect(ref.current).toBe(chip);
    expect(chip).toMatchSnapshot();

    rerender(
      <Chip
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(chip).toMatchSnapshot();

    rerender(<Chip {...props} disabled />);
    expect(chip).toMatchSnapshot();

    rerender(<Chip {...props} theme="outline" />);
    expect(chip).toMatchSnapshot();

    rerender(<Chip {...props} disabled theme="outline" />);
    expect(chip).toMatchSnapshot();
  });

  it("should apply the pressed class name while raisable", () => {
    const modeMock = jest
      .spyOn(INTERACTION_CONFIG, "mode", "get")
      .mockReturnValue("press");
    const { rerender } = render(<Chip raisable>Chip</Chip>);
    const chip = screen.getByRole("button");

    fireEvent.mouseDown(chip);
    expect(chip).toHaveClass("rmd-chip--pressed");
    expect(chip).toMatchSnapshot();

    fireEvent.mouseUp(chip);
    expect(chip).not.toHaveClass("rmd-chip--pressed");
    expect(chip).toMatchSnapshot();

    rerender(<Chip>Chip</Chip>);
    fireEvent.mouseDown(chip);
    expect(chip).not.toHaveClass("rmd-chip--pressed");
    expect(chip).toMatchSnapshot();

    fireEvent.mouseUp(chip);
    expect(chip).not.toHaveClass("rmd-chip--pressed");
    expect(chip).toMatchSnapshot();

    modeMock.mockRestore();
  });

  it("should wrap the children in an additional span unless the disableContentWrap prop is enabled", () => {
    const ref = createRef<HTMLSpanElement>();
    const props = {
      contentStyle: { color: "red" },
      contentClassName: "content-class-name",
      contentProps: {
        ref,
        className: "custom-class-name",
      },
      children: "Content",
    } as const;

    const { rerender } = render(<Chip {...props} />);
    const chip = screen.getByRole("button", { name: "Content" });

    expect(ref.current).toBeInstanceOf(HTMLSpanElement);
    expect(ref.current).toBe(chip.firstElementChild);
    expect(chip).toMatchSnapshot();

    rerender(<Chip {...props} disableContentWrap />);
    expect(ref.current).toBe(null);
    expect(chip).toMatchSnapshot();
  });

  it("should allow for addons to appear before and after the children", () => {
    render(
      <Chip
        leftAddon={<FontIcon>favorite</FontIcon>}
        rightAddon={<FontIcon>close</FontIcon>}
      >
        Content
      </Chip>
    );

    const chip = screen.getByRole("button", { name: "Content" });
    expect(chip).toMatchSnapshot();
  });

  it("should render as a span when the noninteractive prop is enabled", () => {
    const ref = createRef<HTMLButtonElement>();
    const props = {
      "data-testid": "chip",
      ref,
      children: "Content",
    } as const;
    const { rerender } = render(<Chip {...props} noninteractive />);

    let chip = screen.getByTestId("chip");
    expect(chip).toBeInstanceOf(HTMLSpanElement);
    expect(chip).not.toHaveAttribute("aria-pressed");
    expect(chip).not.toHaveAttribute("type");
    expect(chip).toBeEnabled();
    expect(chip).not.toHaveClass("rmd-interaction-surface");
    expect(chip).toMatchSnapshot();

    rerender(<Chip {...props} />);
    chip = screen.getByTestId("chip");
    expect(chip).toBeInstanceOf(HTMLButtonElement);
    expect(chip).toHaveAttribute("type", "button");
    expect(chip).toBeEnabled();
    expect(chip).toHaveClass("rmd-interaction-surface");
    expect(chip).toMatchSnapshot();
  });

  describe("selectable chips", () => {
    it("should update the aria-pressed state based on the selected state for interactable chips", () => {
      const { rerender } = render(
        <Chip data-testid="chip" selected={false}>
          Chip
        </Chip>
      );
      const buttonChip = screen.getByTestId("chip");
      expect(buttonChip).not.toHaveAttribute("aria-pressed");

      rerender(
        <Chip data-testid="chip" selected>
          Chip
        </Chip>
      );
      expect(buttonChip).toHaveAttribute("aria-pressed", "true");

      rerender(
        <Chip data-testid="chip" selected noninteractive>
          Chip
        </Chip>
      );

      const spanChip = screen.getByTestId("chip");
      expect(spanChip).not.toHaveAttribute("aria-pressed", "true");
    });

    it("should apply the selected icon when the selectedThemed prop is not true", () => {
      const props = {
        selected: false,
        children: "Chip",
      } as const;

      const { rerender } = render(<Chip {...props} />);
      const chip = screen.getByRole("button");
      expect(chip).toMatchSnapshot();

      rerender(<Chip {...props} selected />);
      expect(chip).toMatchSnapshot();

      rerender(<Chip {...props} leftAddon={<FontIcon>favorite</FontIcon>} />);
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip {...props} selected leftAddon={<FontIcon>favorite</FontIcon>} />
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip
          {...props}
          leftAddon={<FontIcon>favorite</FontIcon>}
          selectedIconAfter
        />
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip
          {...props}
          selected
          leftAddon={<FontIcon>favorite</FontIcon>}
          selectedIconAfter
        />
      );
      expect(chip).toMatchSnapshot();

      rerender(<Chip {...props} rightAddon={<FontIcon>favorite</FontIcon>} />);
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip {...props} selected rightAddon={<FontIcon>favorite</FontIcon>} />
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip
          {...props}
          rightAddon={<FontIcon>favorite</FontIcon>}
          selectedIconAfter
        />
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip
          {...props}
          selected
          rightAddon={<FontIcon>favorite</FontIcon>}
          selectedIconAfter
        />
      );
      expect(chip).toMatchSnapshot();
    });

    it("should allow for a custom selectedIcon", () => {
      const { rerender } = render(
        <Chip selected={false} selectedIcon={<FontIcon>favorite</FontIcon>}>
          Chip
        </Chip>
      );

      const chip = screen.getByRole("button");
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip selected selectedIcon={<FontIcon>favorite</FontIcon>}>
          Chip
        </Chip>
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip selected={false} selectedIcon="Text">
          Chip
        </Chip>
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip selected selectedIcon="Text">
          Chip
        </Chip>
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip selected={false} selectedIcon="Text" selectedIconAfter>
          Chip
        </Chip>
      );
      expect(chip).toMatchSnapshot();

      rerender(
        <Chip selected selectedIcon="Text" selectedIconAfter>
          Chip
        </Chip>
      );
      expect(chip).toMatchSnapshot();
    });

    it(`should apply ${DISPLAY_NONE_CLASS} to the icon if the disableIconTransition prop is enabled`, () => {
      const props = {
        selected: false,
        children: "Chip",
        selectedThemed: false,
        disableIconTransition: true,
      } as const;

      const { rerender } = render(<Chip {...props} />);
      const chip = screen.getByRole("button");
      expect(chip).toMatchSnapshot();

      rerender(<Chip {...props} selected />);
      expect(chip).toMatchSnapshot();

      rerender(<Chip {...props} selectedIconAfter />);
      expect(chip).toMatchSnapshot();

      rerender(<Chip {...props} selected selectedIconAfter />);
      expect(chip).toMatchSnapshot();
    });

    it("should allow a className to only be applied while selected", async () => {
      const user = userEvent.setup();
      function Test(): ReactElement {
        const { toggled, toggle } = useToggle();
        return (
          <Chip
            selectedClassName="selected-class-name"
            selected={toggled}
            onClick={toggle}
          >
            Chip
          </Chip>
        );
      }

      rmdRender(<Test />);
      const chip = screen.getByRole("button", { name: "Chip" });
      expect(chip).not.toHaveClass("selected-class-name");

      await user.click(chip);
      expect(chip).toHaveClass("selected-class-name");
    });
  });
});
