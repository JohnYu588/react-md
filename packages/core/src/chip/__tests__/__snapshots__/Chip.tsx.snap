// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip selectable chips should allow for a custom selectedIcon 1`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should allow for a custom selectedIcon 2`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--left-addon"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon rmd-chip__icon--visible"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should allow for a custom selectedIcon 3`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  Text
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should allow for a custom selectedIcon 4`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--left-addon"
  type="button"
>
  Text
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should allow for a custom selectedIcon 5`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  Text
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should allow for a custom selectedIcon 6`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--right-addon"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  Text
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the hidden attribute if the disableIconTransition prop is enabled 1`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the hidden attribute if the disableIconTransition prop is enabled 2`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--themed rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the hidden attribute if the disableIconTransition prop is enabled 3`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the hidden attribute if the disableIconTransition prop is enabled 4`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--themed rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 1`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon"
  >
    check
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 2`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--left-addon"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon rmd-chip__icon--visible"
  >
    check
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 3`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 4`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--left-addon"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 5`] = `
<button
  class="rmd-chip rmd-chip--solid rmd-chip--left-addon"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon"
  >
    check
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 6`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--surrounded"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon rmd-chip__icon--visible"
  >
    check
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 7`] = `
<button
  class="rmd-chip rmd-chip--solid rmd-chip--right-addon"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon"
  >
    check
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 8`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--surrounded"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-chip__icon rmd-chip__icon--visible"
  >
    check
  </i>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 9`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip selectable chips should apply the selected icon when the selectedThemed prop is not true 10`] = `
<button
  aria-pressed="true"
  class="rmd-chip rmd-chip--selected rmd-chip--solid rmd-chip--right-addon"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should allow for addons to appear before and after the children 1`] = `
<button
  class="rmd-chip rmd-chip--solid rmd-chip--surrounded"
  type="button"
>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    favorite
  </i>
  <span
    class="rmd-chip__content"
  >
    Content
  </span>
  <i
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    close
  </i>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<button
  class="rmd-chip rmd-chip--solid"
  style="color: white;"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<button
  class="rmd-chip rmd-chip--solid rmd-chip--disabled"
  disabled=""
  style=""
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should apply the correct styling, HTML attributes, and allow a ref 4`] = `
<button
  class="rmd-chip rmd-chip--outline"
  style=""
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should apply the correct styling, HTML attributes, and allow a ref 5`] = `
<button
  class="rmd-chip rmd-chip--outline rmd-chip--disabled"
  disabled=""
  style=""
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should apply the pressed class name while raisable 1`] = `
<button
  class="rmd-chip rmd-chip--solid rmd-chip--pressed rmd-pressed"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
</button>
`;

exports[`Chip should apply the pressed class name while raisable 2`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
</button>
`;

exports[`Chip should apply the pressed class name while raisable 3`] = `
<button
  class="rmd-chip rmd-chip--solid rmd-pressed"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
</button>
`;

exports[`Chip should apply the pressed class name while raisable 4`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Chip
  </span>
</button>
`;

exports[`Chip should render as a span when the noninteractive prop is enabled 1`] = `
<span
  class="rmd-chip rmd-chip--solid rmd-chip--noninteractive"
  data-testid="chip"
>
  <span
    class="rmd-chip__content"
  >
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</span>
`;

exports[`Chip should render as a span when the noninteractive prop is enabled 2`] = `
<button
  class="rmd-chip rmd-chip--solid"
  data-testid="chip"
  type="button"
>
  <span
    class="rmd-chip__content"
  >
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should wrap the children in an additional span unless the disableContentWrap prop is enabled 1`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    class="rmd-chip__content content-class-name custom-class-name"
    style="color: red;"
  >
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;

exports[`Chip should wrap the children in an additional span unless the disableContentWrap prop is enabled 2`] = `
<button
  class="rmd-chip rmd-chip--solid"
  type="button"
>
  <span
    style=""
  >
    Content
  </span>
  <span
    class="rmd-ripple-container"
  />
</button>
`;
